### PR TITLE
New availability cancels shadowed ones

### PR DESCRIPTION
**Pull request by nesteves**
Implements the functionality requested in issue #449 

A user of the application that has a registered phone number can send a text message to the application's number indicating their availability with one of the following formats:

```
available event01

unavailable event01

available custom 2017-12-21 10:00:00 2017-12-21 18:00:00

unavailable custom 2017-12-21 10:00:00 2017-12-21 18:00:00
```

Any one of these will generate at least one availability in the system, with a number of side-effects, described below:

* if the new availability *contains* any existing ones, cancel those, no matter their status
* if the new availability *overlaps* any existing ones, that is, covers a part of their duration
  * having different statuses, move the existing one's boundary so they become contiguous
  * having the same status, cancel the existing one and expand the new one so it covers both
* if the new availability *is contained* by an existing one
  * having different statuses, break the old availability into 2 and accommodate the new one
  * having the same status, return an error message (no availability created)

----
This pull request adds the following to the codebase:

* a series of scopes on the `availability` model that make it easier to do range comparisons
* a validation that prevents the creation of new availabilities that overlap any already existing ones that belong to the targeted person and happen to be active
* a suite of feature specs for receiving availability messages, covering most cases
* a service object for the creation of new availabilities that manipulates existing availabilites to "make room" for the new one

### CODE REVIEW

**Comment by nesteves** [comment]
One thing I remembered, when deployed to production, if there are any active and overlapping (thus invalid) availabilities for a person, she won't be able to message her availability for any periods where these same availabilities take place, as the ensuing updates will fail on validation.

While this might be an issue at the beginning, it will basically fix itself by letting time go by, unless availabilities are defined well in advance.

Any thoughts?

**Comment by kgf** [comment]
I haven't even looked at the code yet, and this is already an impressive PR. The write-up at the beginning is great, thanks.
The only overlap is when people put in for vacation. I'm not too concerned. We'll watch for it.

File: app/models/availability.rb [hunk 1]
```diff
@@ -17,11 +17,32 @@ class Availability < ActiveRecord::Base
...
   scope :available, -> { where(status: "Available")}
   scope :unavailable, -> { where(status: "Unavailable")}
+
   scope :in_the_past, -> { where("start_time <= ?", Time.zone.now) }
-  scope :active, -> { where(status: ['Available', 'Unavailable']).joins(:person).where(people: {status: "Active"}) }
+
+  # These scopes tie the application to postgres, as they rely
+  # on its range data type implementation. The ranges are left_bound closed
+  # and right_bound open, as specified by the argument '[)'
+  # reference: https://www.postgresql.org/docs/9.5/static/rangetypes.html
```
Comment (kgf) [inline app/models/availability.rb:29]: Thanks, I like the comment to more documentation

**Comment by kgf** [comment]
Thanks @nesteves That was a great PR ! Lots of tests, interesting approach, and I'm learning about the PG time stuff.

Thanks for your ghelp,
Kevin

**Comment by nesteves** [comment]
Thank you! I learned a lot too. 

Regards,
Nuno

**Pull request merged**